Default Avatar Image for Users with unset images

salman-neslit opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: #9506

## Description of Changes
- Added default profile avatar images for users.
- Added logic to check if the user's profile contains an `avatarUrl`, if not, a default avatar image will be used.

## "How We Fixed It"
<!-- Brief description of solution, if bug, to be added once issue is resolved. -->
- To ensure users have a profile image, we introduced a check within the profile retrieval logic. If the `avatarUrl` field is not set, the application will default to one of the newly added avatar images.

## Test Plan
- 

## Other Considerations
<!--- Follow-up tickets, breaking changes, etc -->
-

dillchen commented:
Different issue @salman-neslit ?

masvelio commented on packages/commonwealth/client/scripts/utils/avatarUtils.ts:
```diff
@@ -0,0 +1,20 @@
+];
+
+export const getRandomAvatar = (): string => {
+  const randomIndex: number = Math.floor(Math.random() * defaultAvatars.length);
```
default avatar will change on page rerenders/refresh, is this expected behaviour from the product side?

masvelio reviewed (changes requested):
sizes of those images are enormous. Number 5 has almost 700KB! Please pipe them through some optimisation tool that will reduce size and dimension of those avatars so they do not exceed 10kb.

mzparacha commented on packages/commonwealth/client/scripts/utils/avatarUtils.ts:
```diff
@@ -0,0 +1,20 @@
+import avatar5 from './../../assets/img/avatars/default-avatar5.png';
+import avatar6 from './../../assets/img/avatars/default-avatar6.png';
+
+const defaultAvatars: string[] = [
+  avatar1,
+  avatar2,
+  avatar3,
+  avatar4,
+  avatar5,
+  avatar6,
+];
```
I think we should upload this to s3 and then save it in user profile as they'd get used later on, reference

> Additional Context: when we send data to Connect routes for email notifications  #9415, we should send along the preset image profile (assuming the user has not set a profile

from https://github.com/hicommonwealth/commonwealth/issues/9506

Israellund replied:
@mzparacha @salman-neslit Messaged Sachin about this, waiting to hear back. I say wait until we get a definitive answer from Product before doing more work on this. I assume using an s3 will be his answer, but I think it's safer to wait and see.

Israellund replied:
@salman-neslit @mzparacha Just heard from Sachin, the way that this should function is that they should persist unless the user changes the profile image. So, assigned at random and then persisting unless the user changes it

mzparacha replied:
Thanks @Israellund, so @salman-neslit you can upload the new image presets once to s3 (prob look into [uploadImage](https://github.com/hicommonwealth/commonwealth/blob/master/packages/commonwealth/client/scripts/views/components/component_kit/cw_cover_image_uploader.tsx#L137) code and upload via that call) and store the urls in the util, then update user profile with that randomly selected avatar url.

timolegros commented:
@sachben91 @mzparacha @salman-neslit Rather than the proposed solution to save the new avatar when a user who doesn't have one set signs in, could we create a one-time migration to randomly assign avatars to _all_ users who don't have one + modify user creation to also save an avatar URL i.e. make avatar URLs required? This would hugely simplify avatar handling in notifications and emails.

Happy to take over the migration side of this if required.

mzparacha commented:
> @sachben91 @mzparacha @salman-neslit @Israellund Rather than the proposed solution to save the new avatar when a user who doesn't have one set signs in, could we create a one-time migration to randomly assign avatars to all users who don't have one + modify user creation to also save an avatar URL i.e. make avatar URLs required? This would hugely simplify avatar handling in notifications and emails.

I like this better.

sachben91 commented:
Yeah agree on the above

On Fri, Oct 18, 2024 at 7:24 AM Malik Zulqurnain ***@***.***>
wrote:

> @sachben91 <https://github.com/sachben91> @mzparacha
> <https://github.com/mzparacha> @salman-neslit
> <https://github.com/salman-neslit> @Israellund
> <https://github.com/Israellund> Rather than the proposed solution to save
> the new avatar when a user who doesn't have one set signs in, could we
> create a one-time migration to randomly assign avatars to all users who
> don't have one + modify user creation to also save an avatar URL i.e. make
> avatar URLs required? This would hugely simplify avatar handling in
> notifications and emails.
>
> I like this better.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hicommonwealth/commonwealth/pull/9570#issuecomment-2422353782>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJHXV2SXSJOZKEVRTHRYUM3Z4D4WLAVCNFSM6AAAAABQCG2VOCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMRSGM2TGNZYGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Sachin Benny
Newsletter <https://summerlightning.substack.com/>
Farcaster  <https://warpcast.com/sach>

Israellund commented:
@timolegros Yeah I think this is a much better solution.

salman-neslit commented:
> @sachben91 @mzparacha @salman-neslit @Israellund Rather than the proposed solution to save the new avatar when a user who doesn't have one set signs in, could we create a one-time migration to randomly assign avatars to _all_ users who don't have one + modify user creation to also save an avatar URL i.e. make avatar URLs required? This would hugely simplify avatar handling in notifications and emails.
> 
> Happy to take over the migration side of this if required.

@timolegros Thanks for the suggestion! It sounds like a solid approach to simplify avatar handling.I have handled the profile of new users. Please take a look at the code and assist with the migration process.

timolegros commented:
@salman-neslit I added the migration but I also modified how the avatar URL is set when creating a user. Instead of explicitly needing to provide a default URL on `user.create` I added hooks to the model which will populate it automatically. I think this approach is less error prone if we use `user.create` elsewhere in the future.

mzparacha reviewed (approved):
LGTM.

salman-neslit commented:
> @salman-neslit I added the migration but I also modified how the avatar URL is set when creating a user. Instead of explicitly needing to provide a default URL on `user.create` I added hooks to the model which will populate it automatically. I think this approach is less error prone if we use `user.create` elsewhere in the future.

Agreed, this approach seems more reliable
